### PR TITLE
[CALCITE-5807] Add SUBSTRING_INDEX function (enabled in Spark library)

### PR DESCRIPTION
**Pull request by hujianhong**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 1]
```diff
@@ -4179,6 +4179,55 @@ public static Map strToMap(String string, String stringDelimiter, String keyValu
     return map;
   }
 
+  /** Support the SUBSTRING_INDEX function. */
+  public static String substringIndex(String string, String delimiter, int count) {
```
Comment (hujianhong) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:4183]: The implementation reference： https://github.com/apache/spark/blob/f7e8da3fb6dba44ab339a095c46cc8872d83b741/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L819

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 2]
```diff
@@ -6178,6 +6178,34 @@ private static void checkIf(SqlOperatorFixture f) {
...
+    f.checkString("substring_index('www.apache.org', '.', 2)",
+        "www.apache", "VARCHAR(14) NOT NULL");
+    f.checkString("substring_index('www.apache.org', '.', 1)",
+        "www", "VARCHAR(14) NOT NULL");
+    f.checkString("substring_index('www.apache.org', '.', 3)",
+        "www.apache.org", "VARCHAR(14) NOT NULL");
+    f.checkString("substring_index('www.apache.org', '.', -1)",
+        "org", "VARCHAR(14) NOT NULL");
+
+    f.checkNull("substring_index(cast(null as varchar(1)),"
```
Comment (ILuffZhe) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:6199]: We need test cases for those two senarios:

1. `count` is null while `str` and `delimeter` is not null
2. `delimeter` is null but `str` and `count` is not null

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 3]
```diff
@@ -1260,6 +1260,14 @@ private static RelDataType deriveTypeMapFromEntries(SqlOperatorBinding opBinding
           ReturnTypes.IDENTITY_TO_MAP_NULLABLE,
           OperandTypes.STRING_OPTIONAL_STRING_OPTIONAL_STRING);
 
+  /** The "SUBSTRING_INDEX(string, delimiter, count)" function. */
+  @LibraryOperator(libraries = {SPARK})
+  public static final SqlFunction SUBSTRING_INDEX =
+      SqlBasicFunction.create(SqlKind.SUBSTRING_INDEX,
+          ReturnTypes.ARG0_NULLABLE_VARYING,
```
Comment (ILuffZhe) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1267]: If any of the argument is null, this returns null. 
I'm not very familiar with Spark, does this behave like Spark?

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 4]
```diff
@@ -6178,6 +6178,34 @@ private static void checkIf(SqlOperatorFixture f) {
     f.checkNull("str_to_map('a:1,b:2,c:3', ',',null)");
   }
 
+  /** Tests {@code SUBSTRING_INDEX} function from Spark. */
+  @Test void testSubstringIndexFunc() {
```
Comment (ILuffZhe) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:6182]: This function performs a case-sensitive match when searching for delim. I think we should add a test to make a comparison.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 5]
```diff
@@ -6178,6 +6178,34 @@ private static void checkIf(SqlOperatorFixture f) {
     f.checkNull("str_to_map('a:1,b:2,c:3', ',',null)");
   }
 
+  /** Tests {@code SUBSTRING_INDEX} function from Spark. */
```
Comment (ILuffZhe) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:6181]: You can take a look at other java docs in this file. We normally use format like:
`/** Test case for...`.
IMO, it helps others to locate the corresponding JIRA ticket quickly.

**Review by ILuffZhe — COMMENTED** [review]
Left some comments. Please take a look when you are free.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 5 above)
Comment (hujianhong) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:6181]: Yes, I refer to other test cases in this in this file., but their format is ```/** Tests {@code methodName} function from Spark. */```, for example :
```
  /** Tests {@code STR_TO_MAP} function from Spark. */
  @Test void testStrToMapFunc() {
    final SqlOperatorFixture f0 = fixture();
    f0.setFor(SqlLibraryOperators.STR_TO_MAP);
    f0.checkFails("^str_to_map('a=1,b=2', ',', '=')^",
        "No match found for function signature STR_TO_MAP\\("
            + "<CHARACTER>, <CHARACTER>, <CHARACTER>\\)", false);

    final SqlOperatorFixture f = f0.withLibrary(SqlLibrary.SPARK);
    f.checkScalar("str_to_map('a=1,b=2', ',', '=')", "{a=1, b=2}",
```

 Not ```/** Test case for...```.  Is there have a clear rule for java docs in Apache Calcite?

Comment (ILuffZhe) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:6181]: No specific rules for it. As long as it's clear and helps others to find the corresponding ticket.
What do you think of the following example?
```
/** Test case for
   * <a href="https://issues.apache.org/jira/browse/CALCITE-4861">[CALCITE-4861]
   * Optimisation of chained cast calls can lead to unexpected behaviour.</a>.
   */
```

Comment (hujianhong) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:6181]: tks

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 3 above)
Comment (hujianhong) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1267]: Yes，I ran some cases on spark and confirmed that if any of the argument is null, this returns null.

```
select substring_index('abc',cast(null as varchar(1)), cast(null as integer));
+--------------------------------+
| substring_index(abc, CAST(NULL |
| AS STRING), CAST(NULL AS INT)) |
+--------------------------------+
| null                           |
+--------------------------------+


 select substring_index('abc',cast(null as varchar(1)), 1);
+--------------------------------+
| substring_index(abc, CAST(NULL |
|         AS STRING), 1)         |
+--------------------------------+
| null                           |
+--------------------------------+


select substring_index(cast(null as varchar(3)),'.', 1);
+--------------------------------+
|  substring_index(CAST(NULL AS  |
|         STRING), ., 1)         |
+--------------------------------+
| null                           |
+--------------------------------+
```

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 2 above)
Comment (hujianhong) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:6199]: done

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 4 above)
Comment (hujianhong) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:6182]: OK, I will add two cases to make a comparsion.
```
f.checkString("substring_index('aBc', 'B', -1)",
        "c", "VARCHAR(3) NOT NULL");
    f.checkString("substring_index('aBc', 'b', -1)",
        "aBc", "VARCHAR(3) NOT NULL");
```

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 3 above)
Comment (ILuffZhe) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1267]: Thanks for your explanation.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 6]
```diff
@@ -6224,6 +6224,44 @@ private static void checkIf(SqlOperatorFixture f) {
...
+    f.checkNull("substring_index(cast(null as varchar(1)),"
+        + " cast(null as varchar(1)), cast(null as integer))");
+    f.checkNull("substring_index(cast(null as varchar(1)),"
+        + " cast(null as varchar(1)), 2)");
+    f.checkNull("substring_index('abc', cast(null as varchar(1)),"
+        + " cast(null as integer))");
+    f.checkNull("substring_index(cast(null as varchar(1)), '.',"
+        + " cast(null as integer))");
+    f.checkNull("substring_index('abc', '.', cast(null as integer))");
+    f.checkNull("substring_index('abc', cast(null as varchar(1)), 2");
```
Comment (ILuffZhe) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:6262]: You missed one right paren for `substring_index` function. It's better to make your local build pass before pushing.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 7]
```diff
@@ -4206,6 +4206,55 @@ public static Map strToMap(String string, String stringDelimiter, String keyValu
...
+    if (string.isEmpty() || count == 0) {
```
Comment (ILuffZhe) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:4211]: We should also add a test when the count argument is 0.

**Review by ILuffZhe — CHANGES_REQUESTED** [review]
Hi @hujianhong , left some comments.
One small tip: before the final squash and rebase, please don't force push your commit. It helps other reviewers to see the differences.

**Comment by hujianhong** [comment]
> Hi @hujianhong , left some comments. One small tip: before the final squash and rebase, please don't force push your commit. It helps other reviewers to see the differences.

I'm sorry. This is my first PR for calcite. I will be careful in the future.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 7 above)
Comment (hujianhong) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:4211]: done

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 6 above)
Comment (hujianhong) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: done.

**Comment by ILuffZhe** [comment]
There are some specific rules for commit message after https://github.com/apache/calcite/commit/c0e6ba264b3e522010a982ed35de9f3dd03be6af. Could you please amend yours so I can approve the following CI?

`[invalid git log message 'fix and add test cases'; Message must start with upper-case letter]`

**Comment by hujianhong** [comment]
> There are some specific rules for commit message after [c0e6ba2](https://github.com/apache/calcite/commit/c0e6ba264b3e522010a982ed35de9f3dd03be6af). Could you please amend yours so I can approve the following CI?
> 
> `[invalid git log message 'fix and add test cases'; Message must start with upper-case letter]`

done

**Comment by hujianhong** [comment]
@ILuffZhe Can you help me find out what is the reason of the CI task failed? The CI log information seem not match to my PR.
image

**Comment by NobiGo** [comment]
@hujianhong The CI task failure was caused by another PR and has been resolved. So don't worry.

**Comment by ILuffZhe** [comment]
@hujianhong You can refer to the Calcite Website(https://calcite.apache.org/develop/#contributing).
`In the special case, that the CI build failed, and the failure is not caused by your changes create an empty commit (git commit --allow-empty) and push it.`

**Review by ILuffZhe — APPROVED** [review]
LGTM! Let's see if there are any new comments come.

**Comment by ILuffZhe** [comment]
Hi @hujianhong , could you please rebase and squash your commits? I'll merge this in the next few days.

File: site/_docs/reference.md [hunk 8]
```diff
@@ -2764,6 +2764,7 @@ BigQuery's type system uses confusingly different names for types and functions:
 | s | MAP_FROM_ARRAYS(array1, array2)                | Returns a map created from an *array1* and *array2*. Note that the lengths of two arrays should be the same and calcite is using the LAST_WIN strategy
 | s | MAP_FROM_ENTRIES(arrayOfRows)                  | Returns a map created from an arrays of row with two fields. Note that the number of fields in a row must be 2. Note that calcite is using the LAST_WIN strategy
 | s | STR_TO_MAP(string [, stringDelimiter [, keyValueDelimiter]]) | Returns a map after splitting the *string* into key/value pairs using delimiters. Default delimiters are ',' for *stringDelimiter* and ':' for *keyValueDelimiter*. Note that calcite is using the LAST_WIN strategy
+| s | SUBSTRING_INDEX(string, delim, count)          | Returns the substring from **string** before **count** occurrences of the delimiter **delim**. If **count** is positive, everything to the left of the final delimiter (counting from the left) is returned. If **count** is negative, everything to the right of the final delimiter (counting from the right) is returned. The function substring_index performs a case-sensitive match when searching for **delim**.
```
Comment (JiajunBernoulli) [inline site/_docs/reference.md:2767]: Why the `*` is double?

Other always `*arg*`.

Comment (hujianhong) [reply site/_docs/reference.md:2767]: Just referred from spark docucment: https://spark.apache.org/docs/latest/api/sql/index.html#substring_index.

I will fix it.

Comment (hujianhong) [reply site/_docs/reference.md]: I fixed it, please take a look at again, thank you.

**Comment by mihaibudiu** [comment]
@hujianhong if you can squash the two commits we can merge this PR. Thank you.

**Comment by mihaibudiu** [comment]
Can you please rebase and squash the commits so we can merge this PR?

**Comment by hujianhong** [comment]
> Can you please rebase and squash the commits so we can merge this PR?

Ok, I will rebase and squash.

**Comment by mihaibudiu** [comment]
The CI is unhappy, but I think there are only small changes necessary to make it pass.
If you do them, you can probably squash the commits directly.

**Comment by hujianhong** [comment]
> The CI is unhappy, but I think there are only small changes necessary to make it pass. If you do them, you can probably squash the commits directly.

Done, thanks.

**Pull request merged**